pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,182 @@
+// of its size. In latter case, memory is not requested from underlying budget,
+// but we won't be able to do another Get until this message is returned. To
+// track such events, feedBudget uses a counter metric.
+//
+// timeout argument indicates how long we could wait for the budget to replenish
+// if there is not enough budget available. Budget is replenished by Return
+// calls.
+// timeout value of 0 means wait indefinitely.
+// Whenever we use timeout, we return its wait chan to caller, so it could be
+// used subsequently by the chained wait for other resources thus keeping total
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *feedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration, stoppedC <-chan struct{},
+) (sharedBudgetAllocation, <-chan time.Time, error) {
+	// Function to allocate from budget or if there's nothing there borrow from
+	// emergency ether.
+	tryAlloc := func() (sharedBudgetAllocation, error) {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+		if f.mu.closed {
+			return sharedBudgetAllocation{}, errors.New("budget is already released")
+		}
+		if f.mu.oversize {
+			return sharedBudgetAllocation{}, errors.New("failed to allocate mem budget")
+		}
+		err := f.mu.memBudget.Grow(ctx, amount)
+		if err == nil {
+			return sharedBudgetAllocation{size: amount, refCount: 1}, nil
+		}
+		// If we didn't use any budget yet and not currently sending an oversized
+		// event then we can send one.
+		if f.mu.memBudget.Used() == 0 {
+			f.metrics.RangeFeedOverBudgetEvents.Inc(1)
+			f.mu.oversize = true
+			// We are in override mode, we return 0 size to avoid returning it to
+			// budget later as we didn't borrow anything.
+			return sharedBudgetAllocation{size: 0, refCount: 1}, nil
+		}
+		return sharedBudgetAllocation{}, err
+	}
+
+	alloc, err := tryAlloc()
+	if err == nil {
+		return alloc, nil, nil
+	}
+	// If we failed to allocate immediately, then we need to wait on replenish
+	// channel to track Return calls from consumed events.
+	var timeC <-chan time.Time
+	if timeout > 0 {
+		timeC = time.After(timeout)
+	}
+	for {
+		select {
+		case <-f.replenishC:
+			alloc, err = tryAlloc()
+			if err == nil {
+				return alloc, timeC, nil
+			}
+		case <-timeC:
+			return sharedBudgetAllocation{}, timeC, errors.New("failed to allocate mem budget")
+		case <-stoppedC:
+			// We are already stopped, current allocation is already freed so, do
+			// nothing.
+			return sharedBudgetAllocation{}, timeC, nil
+		}
+	}
+}
+
+// Return returns amount to budget.
+func (f *feedBudget) Return(ctx context.Context, x int64) {
+	f.mu.Lock()
+	if f.mu.closed {
+		f.mu.Unlock()
+		return
+	}
+	f.mu.oversize = false
+	if x > 0 {
+		f.mu.memBudget.Shrink(ctx, x)
+	}
+	f.mu.Unlock()
+	select {
+	case f.replenishC <- struct{}{}:
```

Comment:
Doesn't this have a race condition with `Get()`? Consider this interleaving:

* T1: Get() -> ok
* T2: Get() -> tryAlloc() -> err
* T1: Return() -> replenishC -> default
* T2: select replenishC -> times out even though the budget is empty

Reply:
Replenish channel has a size of 1 so default should leave a token there unless there's one already, so T2 would pick it up. That would only work for single producer. Currently we only send data messages from under raftMuLocked and timestamp related messages from async tasks. Since timestamp messages consume no budget they bypass that and it is safe. If we want to extend it to have "out of order" events that we need to account for, this needs to be changed to work with multiple producer/multiple consumer case, something similar to what buffer does in cdc.

Reply:
Ah, gotcha -- missed that. Yeah, since there's only a single producer I think that should be sufficient. I'll have another look just in case.